### PR TITLE
fix: resolve defaultRegion only if necessary

### PR DESCRIPTION
**Pull request by imabhichow**
*Issue #, if available:*

- https://github.com/aws/aws-encryption-sdk-java/issues/1768

*Description of changes:*

- Defer calling `getSdkDefaultRegion()` only if no default region is provided.
- This change only resolve the default region from `DefaultAwsRegionProviderChain` if default region is null or not configured.
- Furthermore, `_defaultRegion` and `_discoveryMrkRegion` will always have non-null region values during build.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

# Check any applicable:
- [ ] Were any files moved? Moving files changes their URL, which breaks all hyperlinks to the files.

### CODE REVIEW

File: src/test/java/com/amazonaws/encryptionsdk/kms/AwsKmsMrkAwareMasterKeyProviderTest.java [hunk 1]
```diff
@@ -225,33 +224,23 @@ public void no_duplicate_mrks() {
...
+          // it will pick one up from the environment
+          .withDefaultRegion(null)
           .buildStrict("mrk-edb7fe6942894d32ac46dbb1c922d574");
     }
 
     @Test
-    // = compliance/framework/aws-kms/aws-kms-mrk-aware-master-key-provider.txt#2.6
-    // = type=test
-    // # If an AWS SDK Default Region can not be
-    // # obtained initialization MUST fail.
```
[the remark this answers is not in the document]
Comment (imabhichow) [reply src/test/java/com/amazonaws/encryptionsdk/kms/AwsKmsMrkAwareMasterKeyProviderTest.java:244]: The check is present because `KmsMasterKeyProvider` allowed `_defaultregion` to be set to null during the build. This is no longer the case with this change. By deferring the call to `getSdkDefaultRegion()` unless no default region is provided, `_defaultregion` will always be a non-null value unless `DefaultAwsCredentialsProvider` could not fetch the region.

[the remark this answers is not in the document]
Comment (texastony) [reply src/test/java/com/amazonaws/encryptionsdk/kms/AwsKmsMrkAwareMasterKeyProviderTest.java:244]: Wait... 
```
    private static String getSdkDefaultRegion() {
      try {
        return new com.amazonaws.regions.DefaultAwsRegionProviderChain().getRegion();
      } catch (SdkClientException ex) {
        return null;
      }
    }
```
This method literally returns `null` in case of an Exception. 
I think this introduces an NPE...

[the remark this answers is not in the document]
Comment (texastony) [reply src/test/java/com/amazonaws/encryptionsdk/kms/AwsKmsMrkAwareMasterKeyProviderTest.java:244]: No, b/c it was already set to be this way.
Darwin has just moved the call further down the process, as the requester has asked.

**Review by texastony — DISMISSED** [review]

**Comment by texastony** [comment]
@seebees @imabhichow I refactored this to maintain the previous behavior w.r.t. ensuring a region is set by the user IF they pass any KeyIDs without a Region.

@imabhichow your solution broke that behavior,
hence you had to change the unit tests. 

However, the best solution maybe to use the MPL.
We should audit the MPL to see when a Default a Region is loaded.
If the MPL's solution is already optimal,
we can direct the requester of #1768 to use the MPL instead.

This would mean they would not have to wait for a new minor release of the ESDK-Java.

File: src/main/java/com/amazonaws/encryptionsdk/kms/AwsKmsMrkAwareMasterKeyProvider.java [hunk 2]
```diff
@@ -413,6 +423,25 @@ private AwsKmsMrkAwareMasterKeyProvider(
     }
 
     this.regionalClientSupplier_ = supplier;
+    // NPE would be thrown via result of extractRegion passed to KMS Client request around line 583.
```
Comment (seebees) [inline src/main/java/com/amazonaws/encryptionsdk/kms/AwsKmsMrkAwareMasterKeyProvider.java:426]: /NPE/Nitro Enclave/s right?

File: src/test/java/com/amazonaws/encryptionsdk/kms/AwsKmsMrkAwareMasterKeyProviderTest.java [hunk 3]
```diff
@@ -224,23 +225,32 @@ public void no_duplicate_mrks() {
...
+          AwsCryptoException.class,
+          () ->
+              AwsKmsMrkAwareMasterKeyProvider.builder()
+                  .withDefaultRegion(null)
+                  .buildStrict("mrk-edb7fe6942894d32ac46dbb1c922d574"));
       AwsKmsMrkAwareMasterKeyProvider.builder()
-          // If default region is set to `null` or not configured
-          // it will pick one up from the environment
-          .withDefaultRegion(null)
+          .withDefaultRegion("us-east-1")
```
Comment (seebees) [inline src/test/java/com/amazonaws/encryptionsdk/kms/AwsKmsMrkAwareMasterKeyProviderTest.java:235]: I don't understand this test. The name is "always need a region", but a region is passed, so I'm confused.

Comment (seebees) [reply src/test/java/com/amazonaws/encryptionsdk/kms/AwsKmsMrkAwareMasterKeyProviderTest.java:235]: Wait! Is this saying "See, `null` fails and `us-east-1` passes, must be right?

This is putting back the test right?

**Review by seebees — COMMENTED** [review]
If my comment is right, I think this is GTG

**Comment by seebees** [comment]
Also, note: this is not a problem in Keyrings.
https://github.com/awslabs/aws-encryption-sdk-specification/blob/a10e0d8889618853f6960ee5f9d622c6d43bc15e/framework/aws-kms/aws-kms-multi-keyrings.md?plain=1#L77-L79

File: src/main/java/com/amazonaws/encryptionsdk/kms/AwsKmsMrkAwareMasterKeyProvider.java (hunk 2 above)
Comment (texastony) [reply src/main/java/com/amazonaws/encryptionsdk/kms/AwsKmsMrkAwareMasterKeyProvider.java:426]: Null Pointer Exception

File: src/test/java/com/amazonaws/encryptionsdk/kms/AwsKmsMrkAwareMasterKeyProviderTest.java (hunk 3 above)
Comment (texastony) [reply src/test/java/com/amazonaws/encryptionsdk/kms/AwsKmsMrkAwareMasterKeyProviderTest.java]: That is correct.
Because this library has a weird feature where `mrk-edb7fe6942894d32ac46dbb1c922d574` is 
considered a valid ARN if and only if a region is given.

**Review by seebees — APPROVED** [review]
LGTM

File: src/main/java/com/amazonaws/encryptionsdk/kms/AwsKmsMrkAwareMasterKeyProvider.java [hunk 4]
```diff
@@ -193,6 +193,25 @@ public AwsKmsMrkAwareMasterKeyProvider.Builder withClientBuilder(AWSKMSClientBui
...
+        // default region is found. Because a default region _may_ not be needed.
+        // Note: Final Null check is at
+        // com.amazonaws.encryptionsdk.kms.AwsKmsMrkAwareMasterKeyProvider#AwsKmsMrkAwareMasterKeyProvider#L417
+        try {
+          // = compliance/framework/aws-kms/aws-kms-mrk-aware-master-key-provider.txt#2.6
+          // # In
+          // # discovery mode if a default MRK Region is not configured the AWS SDK
+          // # Default Region MUST be used.
+          discoveryMrkRegion_ =
+              new com.amazonaws.regions.DefaultAwsRegionProviderChain().getRegion();
```
Comment (imabhichow) [inline src/main/java/com/amazonaws/encryptionsdk/kms/AwsKmsMrkAwareMasterKeyProvider.java:208]: Since we're directly calling Default Provider Chain, we could remove `getSdkDefaultRegion`
```
private static String getSdkDefaultRegion() {
      try {
        return new com.amazonaws.regions.DefaultAwsRegionProviderChain().getRegion();
      } catch (SdkClientException ex) {
        return null;
      }
    }
```

File: src/test/java/com/amazonaws/encryptionsdk/kmssdkv2/AwsKmsMrkAwareMasterKeyProviderTest.java [hunk 5]
```diff
@@ -256,7 +255,7 @@ public void always_need_a_region() {
     // # obtained initialization MUST fail.
     public void discovery_region_can_not_be_null() {
       assertThrows(
-          IllegalArgumentException.class,
+          AwsCryptoException.class,
```
Comment (imabhichow) [inline src/test/java/com/amazonaws/encryptionsdk/kmssdkv2/AwsKmsMrkAwareMasterKeyProviderTest.java:258]: This test would never throw an exception, even though we set `discoveryMrkRegion` set to `null` in builder. It will try to fetch Default region from Credentials Provider chain with the new changes.

**Review by imabhichow — COMMENTED** [review]
We could also consider these two scenarios.

1. What is the expected behaviour if we don't set a `defaultRegion` in builder? Currently, we would try to resolve to the SDK Default Region. With the new changes, we would no longer attempt to resolve the region, and the `defaultRegion` remains `null`. 
 This is the reason for other test failure
```
AwsKmsMrkAwareMasterKeyProviderTest$getMasterKey.other_basic_uses:436 ? AwsCrypto Can't use non-ARN key identifiers or aliases when no default region is set
```
2. What is the expected behaviour when we set `defaultRegion` to `null` in builder? Currently, `defaultRegion` would remain `null`. With new changes, the behaviour would remain the same.

File: src/test/java/com/amazonaws/encryptionsdk/kmssdkv2/AwsKmsMrkAwareMasterKeyProviderTest.java (hunk 5 above)
Comment (texastony) [reply src/test/java/com/amazonaws/encryptionsdk/kmssdkv2/AwsKmsMrkAwareMasterKeyProviderTest.java:258]: Oh, your right, I went too far.
We should revert this:
https://github.com/aws/aws-encryption-sdk-java/blob/69cfee6cfc9403335925a35fd35717aecce94c7d/src/main/java/com/amazonaws/encryptionsdk/kmssdkv2/AwsKmsMrkAwareMasterKeyProvider.java#L170-L188

**Review by texastony — CHANGES_REQUESTED** [review]
Blocking Merge until this is resolved:
https://github.com/aws/aws-encryption-sdk-java/pull/1964/files#r1515103849

**Comment by texastony** [comment]
> 1. With the new changes, we would no longer attempt to resolve the region, and the `defaultRegion` remains `null`.

I don't believe this correct. I think all end cases set `defaultRegion`, just post the builder, in the `AwsKmsMrkAwareMasterKeyProvider` method.

File: src/main/java/com/amazonaws/encryptionsdk/kmssdkv2/AwsKmsMrkAwareMasterKeyProvider.java [hunk 6]
```diff
@@ -348,6 +352,26 @@ private AwsKmsMrkAwareMasterKeyProvider(
...
+        if (defaultRegion == null) {
+          throw new AwsCryptoException(
+              "Default Region was not set and Default AWS Region Provider Chain could not load one.");
+        }
+      } catch (SdkClientException ex) {
+        throw new AwsCryptoException(
+            "Default Region was not set and Default AWS Region Provider Chain could not load one.",
+            ex);
+      }
+    }
```
Comment (imabhichow) [inline src/main/java/com/amazonaws/encryptionsdk/kmssdkv2/AwsKmsMrkAwareMasterKeyProvider.java:374]: I think we should move this behaviour before we check for non ARN's KeyIds either in Builder or before the check. Otherwise, we would still expect exception being thrown.

Comment (texastony) [reply src/main/java/com/amazonaws/encryptionsdk/kmssdkv2/AwsKmsMrkAwareMasterKeyProvider.java:374]: I do not follow. 
Doesn't this check have to happen AFTER the non ARN Key IDs?
Or the illegal combination of "no region with non ARN" cannot be detected.

I believe the intention is to prevent users from providing non ARN Key IDs if the region is not set.
That is what the Unit tests assert.

Comment (imabhichow) [reply src/main/java/com/amazonaws/encryptionsdk/kmssdkv2/AwsKmsMrkAwareMasterKeyProvider.java:374]: I guess you're correct. 

However, there are [other unit tests](https://github.com/aws/aws-encryption-sdk-java/blob/e94ee8538501f270138feab0a6138bed7ce640ea/src/test/java/com/amazonaws/encryptionsdk/kmssdkv2/AwsKmsMrkAwareMasterKeyProviderTest.java#L489-L507) that check if the default Region is not configured and are passing non-ARN keys. It should resolve with the AWS SDK region. I believe the original changes were asserted such that [defaultRegion == AWS SDK region](https://github.com/aws/aws-encryption-sdk-java/blob/e94ee8538501f270138feab0a6138bed7ce640ea/src/main/java/com/amazonaws/encryptionsdk/kms/AwsKmsMrkAwareMasterKeyProvider.java#L83). As a result, if defaultMrkRegion is not configured, it will resolve to defaultRegion, and if the AWS SDK Region cannot be fetched, we expect a failure.

Comment (texastony) [reply src/main/java/com/amazonaws/encryptionsdk/kmssdkv2/AwsKmsMrkAwareMasterKeyProvider.java:374]: I think this code has evolved to be convoluted.
I do not know what the purpose of these conflicting Unit Tests are trying to yield...

I agree, that the current deployed code behaves as you described it:
"An inferred Region is sufficient to allow for non ARNs", 
at least for STRICT, which is what `other_basic_uses` tests.

I am not sure what the behavior is supposed to be for Discovery,
but I... I guess it is the same: "inferred is sufficient if not provided".

So I was entirely wrong about "no region with non ARN" being an illegal combination.

If we set the region in the Builder, we will need to refactor the unit tests to mock the EC2 data loader to confirm that a null returned there fails the build.
